api/src/main/java/com/nicknathanjustin/streamercontracts/votes/VoteServiceImpl.java

```diff
@@ -0,0 +1,130 @@
+package com.nicknathanjustin.streamercontracts.votes;
+
+import com.nicknathanjustin.streamercontracts.contracts.ContractModel;
+import com.nicknathanjustin.streamercontracts.users.UserModel;
+import lombok.NonNull;
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+
+import java.sql.Timestamp;
+import java.util.UUID;
+
+@RequiredArgsConstructor
+@Slf4j
+public class VoteServiceImpl implements VoteService{
+
+    @NonNull final VoteModelRepository voteModelRepository;
+
+    @Override
+    public void recordVote(@NonNull final UserModel voter,
+                           @NonNull final ContractModel contractModel,
+                           @NonNull final boolean flaggedCompleted) {
+        if (userCanVoteOnContract(voter, contractModel)) {
+            final VoteModel voteModel = VoteModel.builder()
+                    .voter(voter)
+                    .contract(contractModel)
+                    .viewerFlaggedComplete(flaggedCompleted)
+                    .votedAt(new Timestamp(System.currentTimeMillis()))
+                    .build();
+            voteModelRepository.save(voteModel);
+        }
+    }
+
+    @Override
+    public boolean isVotingComplete(@NonNull final ContractModel contractModel) {
+        validateContract(contractModel);
+
+        final UUID contractId = contractModel.getId();
+        final VoteModel proposerVote = voteModelRepository.findByContractIdAndVoterId(contractId, contractModel.getProposer().getId());
+        final VoteModel streamerVote = voteModelRepository.findByContractIdAndVoterId(contractId, contractModel.getStreamer().getId());
+        final Timestamp now = new Timestamp(System.currentTimeMillis());
+        return (proposerVote != null && streamerVote != null) ||
+               (proposerMarkedContractCompleted(contractModel)) ||
+               (streamerMarkedContractFailed(contractModel)) ||
+               (now.after(contractModel.getExpiresAt()));
+    }
+
+    @Override
+    public boolean wasContractCompleted(@NonNull final ContractModel contractModel) {
```

Comment:
I think this function should actually return an enum of 1 of 3 states. COMPLETED, FAILED, or DISPUTE.

Reply:
Good callout. I like that too. Helps clarity. Switched to suggested.